A36-xds-for-servers.md

```diff
@@ -0,0 +1,316 @@
+xDS-powered gRPC features can be used without code changes.
+
+### Related Proposals:
+
+ * [A27: xDS-Based Global Load Balancing][A27]
+ * [A29: xDS-Based Security for gRPC Clients and Servers][A29]
+
+[A27]: A27-xds-global-load-balancing.md
+[A29]: A29-xds-tls-security.md
+
+## Proposal
+
+Each language will create an "XdsServer" API that will wrap the normal "Server"
+API. This will allow it to see the listening ports, start event, stop event, and
+also allow it access to the server to change behavior (e.g., via an
+interceptor). Users will use the XdsServer API to construct the server instead
+of the existing API.
+
+Credential configuration will be managed separately by providing a
+XdsServerCredential, similar to client-side. The user must pass the
+XdsServerCredential to the XdsServer API to enable control-plane managed TLS
+certificates. The user is free to pass other ServerCredential types to XdsServer
+and they will be honored. Note that RBAC is authz, not authn, so it would be
+enabled by using the XdsServer, even without XdsServerCredential.
+
+When starting the XdsServer, it will start communication with xDS to receive
+initial configuration. `bind()` and `listen()` can happen immediately, but
+`accept()`ing on the port will be delayed until the initial configuration is
+received. `listen()` should be delayed if possible, but commonly needs to happen
+immediately after `bind()` for implementation-specific reasons. If initial
+configuration fails, the server will continue delaying and server startup will
+stall until it succeeds. Each language will provide an XdsServer-specific API to
+inform the application of initial communication failures. If the user does not
+configure the API, gRPC should log the errors at a default-visible log level. If
+the xds bootstrap is missing or invalid, implementations would ideally fail
+server startup, but a permanent hang is also acceptable. "Initial configuration"
+does not have to include server credentials.
+
+The `GRPC_XDS_BOOTSTRAP` file will be enhanced to have a new field:
+```json
+{
+  "grpc_server_resource_name_id": "example/resource",
+  // ...
+}
+```
+
+XdsServer will use the normal `XdsClient` to communicate with the xDS server.
+The XdsServer will pass the listening address (commonly using a wildcard IP
+address, like `::` or `0.0.0.0`) to a new XdsClient API to start a watch for a
+`envoy.config.listener.v3.Listener` resource. XdsClient will concatenate the
+value of `grpc_server_resource_name_id` from the bootstrap,
+`?udpa.resource.listening_address=`, and the listening address to construct the
+resource name. For example, with an address of `[::]:80` and the example
+bootstrap field above, the resource name would be
+`example/resource?udpa.resource.listening_address=[::]:80`. Note that the `[]`
+is _not_ percent encoded, as this string is not a URI.
+
+The xDS-returned Listener must have an [`address`][Listener.address] that
```

Comment:
In Go, the piece of code which processes the response from the management server and decides whether to ack or nack, does not know about the set of resources being watched (that happens at a higher layer). This means that when we get a `Listener` resource, we do not have a way to make sure that the `address` and `port_value` in the `SocketAddress` match the one specified in the `listening_address` field. What we *can* do is to make sure that it matches with the `name` field (because eventually the `name` in the response has to match the one in the request).

I wanted to bring this us up to see if there will be any inconsistency in the handling of this across the languages.

Reply:
There is no `listening_address` field. It is just encoded in the resource name in the request. So when you match the name in the response to the name in the request you are indirectly verifying the value of the `listening_address`. In that case it should not be difficult to match that part with the `[address][Listener.address]` value either.

Reply:
Repeating my earlier comment from reviewable:
> Hmm, good question. Maybe the `XdsClient` can verify that the `Listener` resource has its address field set to the same address as in the resource name. Although that might be difficult if we use the template approach I suggest above instead of hard-coding "xds.resource.listener_address=" in our implementation.

We have now switched to the template approach.  Given that, I don't think we can make any assumptions about the `name` field.

I would like to do this validation, but I don't see a reasonable way to do it without some fairly ugly layering violations.  Unless one of you can think of another way to do this, I think we may need to live without this check. :(

Reply:
I can think of the following layered approach for validating this:

- the xDSClient validates the returned Listener by matching the resource name against the requested resource name (trivial common check for all resources?)
- the actual watcher that requested this Listener resource then validates that the returned Listener has the [`address`][Listener.address] matching the expected value. This might be too late to NACK to the control plane but should also be rare (a bug in the control plane).